File: src/sc_neurocore/neurons/models/lnm.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — Fully parameterized learnable neuron (Jolivet et al. 2006 threshold models)

from __future__ import annotations

import math
from dataclasses import dataclass


@dataclass
class LearnableNeuronModel:
    """Fully parameterized learnable neuron.

    V[t+1] = alpha * V[t] + beta * I[t] + gamma * f(V[t])
    where alpha, beta, gamma are trainable scalars and f is a
    learnable activation (here sigmoid) — a trainable generalisation of
    the simple threshold models fitted to cortical recordings by Jolivet et al.

    Reference: Jolivet, Rauch, Lüscher & Gerstner (2006). Predicting spike
    timing of neocortical pyramidal neurons by simple threshold models.
    J Comput Neurosci 21:35-49.
    """

    v: float = 0.0
    alpha: float = 0.9
    beta: float = 0.1
    gamma: float = 0.05
    v_threshold: float = 1.0
    v_reset: float = 0.0
    f_slope: float = 5.0  # sigmoid steepness
    f_shift: float = 0.5  # sigmoid center

    def __post_init__(self) -> None:
        for name in ("v", "alpha", "beta", "gamma", "v_reset", "f_shift"):
            if not math.isfinite(getattr(self, name)):
                raise ValueError(f"{name} must be finite")
        for name in ("v_threshold", "f_slope"):
            value = getattr(self, name)
            if not math.isfinite(value) or value <= 0:
                raise ValueError(f"{name} must be finite and positive")

    @staticmethod
    def _sigmoid(value: float) -> float:
        if value >= 0.0:
            z = math.exp(-value)
            return 1.0 / (1.0 + z)
        z = math.exp(value)
        return z / (1.0 + z)

    def step(self, current: float) -> int:
        if not math.isfinite(current):
            raise ValueError("current must be finite")

        f_v = self._sigmoid(self.f_slope * (self.v - self.f_shift))
        self.v = self.alpha * self.v + self.beta * current + self.gamma * f_v
        if self.v >= self.v_threshold:
            self.v = self.v_reset
            return 1
        return 0

    def reset(self) -> None:
        self.v = 0.0

File: engine/src/neurons/simple_spiking/lnm.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SC-NeuroCore — Learnable Neuron Model

//! Learnable parameterized neuron dynamics.

/// Learnable Neuron Model (LNM) — parameterised activation + decay.
#[derive(Clone, Debug)]
pub struct LearnableNeuronModel {
    pub v: f64,
    pub alpha: f64,
    pub beta: f64,
    pub gamma: f64,
    pub v_threshold: f64,
    pub f_slope: f64,
    pub f_shift: f64,
}

impl LearnableNeuronModel {
    pub fn new() -> Self {
        Self {
            v: 0.0,
            alpha: 0.9,
            beta: 0.1,
            gamma: 0.05,
            v_threshold: 1.0,
            f_slope: 5.0,
            f_shift: 0.5,
        }
    }
    pub fn step(&mut self, current: f64) -> i32 {
        let f_v = 1.0 / (1.0 + (-(self.f_slope * (self.v - self.f_shift))).exp());
        self.v = self.alpha * self.v + self.beta * current + self.gamma * f_v;
        if self.v >= self.v_threshold {
            self.v = 0.0;
            1
        } else {
            0
        }
    }
    pub fn reset(&mut self) {
        self.v = 0.0;
    }
}
impl Default for LearnableNeuronModel {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn default_matches_constructor_state() {
        let default = LearnableNeuronModel::default();
        let constructed = LearnableNeuronModel::new();
        assert_eq!(default.v, constructed.v);
    }

    #[test]
    fn lnm_fires() {
        let mut n = LearnableNeuronModel::new();
        let t: i32 = (0..50).map(|_| n.step(2.0)).sum();
        assert!(t > 0);
    }

    #[test]
    fn lnm_reset_clears_state() {
        let mut n = LearnableNeuronModel::new();
        for _ in 0..50 {
            n.step(2.0);
        }
        n.reset();
        assert!((n.v - 0.0).abs() < 1e-10);
    }

    #[test]
    fn lnm_bounded() {
        let mut n = LearnableNeuronModel::new();
        for _ in 0..1000 {
            n.step(100.0);
        }
        assert!(n.v.is_finite());
    }

    #[test]
    fn lnm_nan_no_panic() {
        LearnableNeuronModel::new().step(f64::NAN);
    }
}
